src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,582 @@ const std::string& TNAuthList::Entry::telephone_number() const {
    return std::get<ASN1_String>(m_data).value();
 }
 
+std::vector<uint8_t> IPAddressBlocks::encode_inner() const {
+   std::vector<uint8_t> output;
+   DER_Encoder(output).start_sequence().encode_list(m_ip_addr_blocks).end_cons();
+   return output;
+}
+
+void IPAddressBlocks::decode_inner(const std::vector<uint8_t>& in) {
+   BER_Decoder(in).decode_list(m_ip_addr_blocks).verify_end();
+}
+
+void IPAddressBlocks::sort_and_merge() {
+   // std::map is ordered, so a uint32_t of [afi,safi] will be in the right order
+   std::map<uint32_t, std::vector<IPAddressFamily>> afam_map;
+   for(IPAddressFamily& block : m_ip_addr_blocks) {
+      uint32_t afam = block.afi();
+      if(block.safi().has_value()) {
+         afam = static_cast<uint32_t>(afam << 8) | block.safi().value();
+      }
+      std::vector<IPAddressFamily>& fams = afam_map[afam];
+      fams.push_back(block);
+   }
+
+   std::vector<IPAddressFamily> merged_blocks;
+   for(auto it = afam_map.begin(); it != afam_map.end(); it++) {
+      std::vector<IPAddressFamily>& fams = it->second;
+
+      // fams[0] has to have the same choice type as the fams in the same bucket
+      if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(fams[0].addr_choice())) {
+         merged_blocks.push_back(merge<Version::IPv4>(fams));
+      } else {
+         merged_blocks.push_back(merge<Version::IPv6>(fams));
+      }
+   }
+   m_ip_addr_blocks = merged_blocks;
+}
+
+template <IPAddressBlocks::Version V>
+IPAddressBlocks::IPAddressFamily IPAddressBlocks::merge(std::vector<IPAddressFamily>& blocks) {
+   if(blocks.size() == 1) {
+      return blocks[0];
+   }
+
+   bool all_inherit = true;
+   bool none_inherit = true;
+   for(IPAddressFamily& block : blocks) {
+      IPAddressChoice<V> choice = std::get<IPAddressChoice<V>>(block.addr_choice());
+      all_inherit = !choice.ranges().has_value() && all_inherit;
+      none_inherit = choice.ranges().has_value() && none_inherit;
+   }
+
+   if(all_inherit) {
+      return IPAddressFamily(IPAddressChoice<V>(), blocks[0].safi());
```

Comment:
If `blocks` is empty this crashes. This should be checked earlier and either handled correctly, if that's possible, or otherwise rejected with an exception.